diff --git a/Makefile.in b/Makefile.in
index 21f33d7..c87d0de 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -660,6 +660,9 @@ stmp-config-host-ppl: config.status stmp-install-host-gmp
 	 mkdir ${PPL_VERSION}) && \
 	($(canadian_env) \
 	 cd ${PPL_VERSION} ; \
+	 find $(call find-package-srcdir, ${PPL_VERSION})/src -name '*.hh' | xargs fgrep -l ptrdiff_t | while read fn; do \
+	 fgrep -q cstddef "$$fn" || sed -i -e 's/ptrdiff_t/std::&/g' -e '1i'\\\\\
+'#include <cstddef>' "$$fn" ; done ; \
          export CFLAGS="${CFLAGS} -I$(GMP_DIR)/include" ; \
          export CXXFLAGS="${CFLAGS} -I$(GMP_DIR)/include" ; \
          export LDFLAGS="${LDFLAGS} -L$(GMP_DIR)/lib" ; \
